x-pack/legacy/plugins/infra/public/containers/logs/log_entries/index.ts

```diff
@@ -0,0 +1,249 @@
+  if (pagesAfterEnd < DESIRED_BUFFER_PAGES && hasMoreAfterEnd) return ShouldFetchMoreEntries.After;
+  return false;
+};
+
+const useFetchEntriesEffect = (
+  state: LogEntriesStateParams,
+  dispatch: Dispatch,
+  params: FetchEntriesParams
+) => {
+  const { getLogEntriesAround, getLogEntriesBefore, getLogEntriesAfter } = useGraphQLQueries();
+
+  const [prevParams, cachePrevParams] = useState(params);
+
+  const runFetchNewEntriesRequest = async () => {
+    dispatch({ type: Action.FetchingNewEntries });
+    try {
+      const payload = await getLogEntriesAround(params);
+      dispatch({ type: Action.ReceiveNewEntries, payload });
+    } catch (e) {
+      dispatch({ type: Action.ErrorOnNewEntries });
+    }
+  };
+
+  const runFetchMoreEntriesRequest = async (direction: ShouldFetchMoreEntries) => {
+    dispatch({ type: Action.FetchingMoreEntries });
+    const getEntriesBefore = direction === ShouldFetchMoreEntries.Before;
+    const timeKey = getEntriesBefore
+      ? state.entries[0].key
+      : state.entries[state.entries.length - 1].key;
+    const getMoreLogEntries = getEntriesBefore ? getLogEntriesBefore : getLogEntriesAfter;
+    try {
+      const payload = await getMoreLogEntries({ ...params, timeKey });
+      dispatch({
+        type: getEntriesBefore ? Action.ReceiveEntriesBefore : Action.ReceiveEntriesAfter,
+        payload,
+      });
+    } catch (e) {
+      dispatch({ type: Action.ErrorOnMoreEntries });
+    }
+  };
+
+  const fetchNewEntriesEffectDependencies = Object.values(
+    pick(params, ['sourceId', 'filterQuery', 'timeKey'])
+  );
+  const fetchNewEntriesEffect = () => {
+    if (shouldFetchNewEntries({ ...params, ...state, prevParams })) {
+      runFetchNewEntriesRequest();
+    }
+    cachePrevParams(params);
+  };
+
+  const fetchMoreEntriesEffectDependencies = [
+    ...Object.values(pick(params, ['pagesAfterEnd', 'pagesBeforeStart'])),
+    Object.values(pick(state, ['hasMoreBeforeStart', 'hasMoreAfterEnd'])),
+  ];
+  const fetchMoreEntriesEffect = () => {
+    if (state.isLoadingMore) return;
+    const direction = shouldFetchMoreEntries(params, state);
+    switch (direction) {
+      case ShouldFetchMoreEntries.Before:
+      case ShouldFetchMoreEntries.After:
+        runFetchMoreEntriesRequest(direction);
+        break;
+      default:
+        break;
+    }
+  };
+  useEffect(fetchNewEntriesEffect, fetchNewEntriesEffectDependencies);
+  useEffect(fetchMoreEntriesEffect, fetchMoreEntriesEffectDependencies);
+
+  return {
+    fetchNewerEntries: useCallback(() => runFetchMoreEntriesRequest(ShouldFetchMoreEntries.After), [
+      params,
+    ]),
+  };
+};
+
+export const useLogEntriesState: (
+  params: LogEntriesFetchParams
+) => [LogEntriesStateParams, LogEntriesCallbacks] = params => {
+  const [state, dispatch] = useReducer(logEntriesStateReducer, logEntriesInitialState);
+
+  const { fetchNewerEntries } = useFetchEntriesEffect(state, dispatch, {
+    ...params,
+    sourceId: 'default',
```

Comment:
I think we need to pass this as a parameter

Reply:
This is much easier to do after https://github.com/elastic/kibana/pull/51526/commits/3b5227da29035ecd6da681e6368c6712c5b9eaaa so I'll take care of that